db/kv_channel_storage_test.go

```diff
@@ -0,0 +1,331 @@
+	latestSequence []uint64       // Tracks high sequence per vbucket for the generator
+	rnd            *rand.Rand     // Random generator
+}
+
+func (l *LogEntryGenerator) Close() {
+	close(l.terminator)
+}
+
+func (l *LogEntryGenerator) Start() {
+	l.Output = make(chan *LogEntry, 20)
+	l.terminator = make(chan bool)
+	l.latestSequence = make([]uint64, 1024)
+	l.rnd = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	go l.WriteEntries()
+}
+
+func (l *LogEntryGenerator) WriteEntries() {
+	for {
+		select {
+		case l.Output <- l.MakeNextEntry():
+		case <-l.terminator:
+			return
+		}
+	}
+}
+
+func (l *LogEntryGenerator) MakeNextEntry() *LogEntry {
+
+	vbNo := rand.Intn(1024)
+	// TODO: random distribution around SequenceGap, instead of fixed step
+	l.latestSequence[vbNo] = l.latestSequence[vbNo] + l.SequenceGap
+
+	entry := LogEntry{
+		VbNo:     uint16(vbNo),
+		Sequence: l.latestSequence[vbNo],
+		DocID:    fmt.Sprintf("generated_document_%d_%d", vbNo, l.latestSequence[vbNo]),
+		RevID:    "1-480a0a76c43f80e572405c164ffc7e3d",
+	}
+
+	return &entry
+}
+
+// ------------------------------------------------
+// ChannelStorageBenchmarkSet - used to manage setup/teardown for tests and benchmarks
+// ------------------------------------------------
+type ChannelStorageBenchmarkSet struct {
+	indexBucket    base.Bucket
+	channelStorage ChannelStorage
+	generator      LogEntryGenerator
+}
+
+func initChannelStorageBenchmarkSet(storageType ChannelStorageType, sequenceGap uint64, statsEnabled bool) ChannelStorageBenchmarkSet {
+
+	indexBucket := testIndexBucket()
+	if statsEnabled {
+		indexBucket = base.NewStatsBucket(indexBucket)
+	}
+
+	channelStorage := testChannelStorage(storageType, indexBucket, "ABC", 64)
+	generator := LogEntryGenerator{
+		SequenceGap: sequenceGap,
+	}
+	generator.Start()
+	return ChannelStorageBenchmarkSet{
+		indexBucket:    indexBucket,
+		channelStorage: channelStorage,
+		generator:      generator,
+	}
+}
+
+func (c ChannelStorageBenchmarkSet) Close() {
+	c.indexBucket.Close()
+	c.generator.Close()
+}
+
+// ------------------------------------------------
+// WriteEntries - using the specified storage and generator, writes [count] entries.  Returns a validationSet containing the list of entries that were written,
+// and a clock representing the high sequence number for each vbucket (stable clock for the entries)
+// ------------------------------------------------
+func WriteEntries(storage ChannelStorage, generator LogEntryGenerator, count int) (validationSet map[string]*LogEntry, stableClock base.SequenceClock, err error) {
+	stableClock = base.NewSequenceClockImpl()
+	validationSet = make(map[string]*LogEntry)
+	// Write  entries to storage, in batches of 100
+	batchSize := 100
+	remaining := count
+	for remaining > 0 {
+		if remaining < batchSize {
+			batchSize = remaining
+		}
+		entrySet := make([]*LogEntry, batchSize)
+		for j := 0; j < batchSize; j++ {
+			entry := <-generator.Output
```

Comment:
Well I do see one advantage of the goroutine -- it can maintain it's sequence state w/o having to do explicit locking/unlocking.

Mostly asking out of curiosity, not suggesting it needs to be changed.